### ISSUE TITLE
Statics 'behaving' weirdly in functions

### ISSUE THREAD

**Issue by terrarier2111** [phil-opp/blog_os#1098]
Hello, i am writing a hobby OS and i got followed the tutorial until the end of the allocator designs post and today i tried to implement a simple shell on top of what i got. For this i wanted to use a Mutex<Shell> in a lazy static (the Shell contains a String and a bool). At the end of the initialization there was a bool i wanted to change inside the Shell, everything worked fine until i tried to query said bool from the keyboard interrupt. The value in the Shell never changed inside the interrupt function, it did however change in the main function. After some debugging i found out that the memory address of the Shell is always the same inside the main function, but it is completely different in the keyboard interrupt function. After some more debugging i found out that this is true not only for the Shell lazy static. My question here really is: is this intended? and what can i do about it?

**Comment by bjorn3** [comment]
What did you exactly try? Setting an AtomicBool? Locking and unlocking a Mutex<bool> in a loop? Or something else?

**Comment by terrarier2111** [comment]
> What did you exactly try? Setting an AtomicBool? Locking and unlocking a Mutex in a loop? Or something else?

i tried setting an atomic bool, locking and unlocking a mutex (without loop) and even using "normal" data - if it helps i can also make the source code accessible (although it looks terrible because of all the debugging i did)

**Comment by terrarier2111** [comment]
This is how the output looks like

![Bildschirmfoto vom 2022-03-27 11-01-05](https://user-images.githubusercontent.com/58695553/160274541-0006658e-1b69-46d6-bfa0-58518afd4b34.png)

**Comment by terrarier2111** [comment]
I noticed (as becomes visible in the gist) that this weirdness not only applies to interrupts but to some (not all) function calls as well - in this case the ones in lib

**Comment by phil-opp** [comment]
I did some debugging on your code and I think I found the issue. I added some print statements to show the pointer and value of the `TEST` static in `main.rs` and `interrupts.rs` and it turns out that the pointers are different between those two files. This means that the compiler creates two separate versions of that static. So the question was, why?

The issue is that you're defining your submodules in both your `lib.rs` and `main.rs`, i.e. you have a `mod shell;` declaration in both of them. This creates two completly separate copies of all your modules, let's call them `main::shell` and `lib::shell`. For code and constants, this does not really matter, but it does for `statics`. So `main::shell::TEST` is a completely different static than `lib::shell::TEST` this way.

The solution is to move all your `mod` statements to your `lib.rs` and add the following `use` statement in `main.rs` instead:

```rust
use LeafOS::{
    allocators, driver, events, filesystem, print, println, process, scheduler, serial,
    serial_print, serial_println, shell, vga_buffer,
};
```

Then everything should work as expected.
